goal_src/jak2/engine/ambient/ambient.gc

```diff
@@ -9,3 +9,537 @@
+                                     (logtest? (-> *sound-iop-info* stream-status s5-2) (stream-status ststatus-six))
+                                     )
+                            (set! v1-43 #f)
+                            (goto cfg-39)
+                            )
+                          )
+                        (set! v1-43 #t)
+                        (label cfg-39)
+                        (and v1-43 (< (- (-> self clock frame-counter) (-> self state-time)) (seconds 120)))
+                        )
+                      )
+                 (and (nonzero? (-> self message-id))
+                      (= (gui-control-method-17 *gui-control* (-> self message-id)) (gui-action playing))
+                      (or (< (- (-> self clock frame-counter) (-> self state-time)) (the-as time-frame (-> self message text-duration)))
+                          (and (logtest? (-> self message flags) 16) (-> self region) (region-method-9 (-> self region) (target-pos 0)))
+                          )
+                      )
+                 (< (- (-> self clock frame-counter) (-> self state-time)) (seconds 0.05))
+                 )
+        (when (and (nonzero? (-> self voice-id))
+                   (not gp-1)
+                   (zero? (gui-control-method-17 *gui-control* (-> self voice-id)))
+                   )
+          (remove-setting! 'music-volume)
+          (remove-setting! 'sfx-volume)
+          (remove-setting! 'dialog-volume)
+          (set! gp-1 #t)
+          )
+        (when (and (nonzero? (-> self message-id))
+                   (= (gui-control-method-17 *gui-control* (-> self message-id)) (gui-action playing))
+                   (not (paused?))
+                   )
+          (if (logtest? (-> self message flags) 32)
+              (seek! (-> self interp) 1.0 (* 0.5 (-> self clock seconds-per-frame)))
+              (seek! (-> self interp) 1.0 (* 4.0 (-> self clock seconds-per-frame)))
+              )
+          (talker-method-17 self)
+          )
+        (suspend)
+        )
+      )
+    (when (and (nonzero? (-> self message-id)) (logtest? (-> self message flags) 96))
+      (while (!= (-> self interp) 0.0)
+        (if (logtest? (-> self message flags) 32)
+            (seek! (-> self interp) 0.0 (* 0.5 (-> self clock seconds-per-frame)))
+            (seek! (-> self interp) 0.0 (* 4.0 (-> self clock seconds-per-frame)))
+            )
+        (talker-method-17 self)
+        (suspend)
+        )
+      )
+    (when (and (logtest? (-> self message flags) 8) (not (-> self save?)))
+      (let ((gp-2 (-> self clock frame-counter)))
+        (until (>= (- (-> self clock frame-counter) gp-2) (seconds 1))
+          (suspend)
+          )
+        )
+      (set! (-> self save?) #t)
+      (auto-save-user)
+      )
+    (go-virtual exit)
+    (none)
+    )
+  )
+
+(defstate exit (talker)
+  :virtual #t
+  :code (the-as (function none :behavior talker) nothing)
+  )
```

Comment:
the decompilation looks good, but to be honest I have 0 idea what's going on in this file.